Add ruby2_keywords to sequencer

gliavallo opened the pull request:
Calling functions with keyword arguments from `sequencer.rb` produces keyword argument deprecation warnings in Ruby 2.7.

`ruby2_keywords` is added to `Sequencer::OperationWrapper::method_missing` function to fix the warning.

Sources:
https://eregon.me/blog/2021/02/13/correct-delegation-in-ruby-2-27-3.html
https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html

Current problems:
1. Tests does not run under Ruby 2.7+ (solved)
2. Should we add ruby2_keywords dependency somewhere (it worked without it when imported gem) (solved)
3. Should minimal version of Ruby bumped? (solved)

gliavallo committed:
Add ruby2_keywords to sequencer

Calling functions with keyword arguments from `sequencer.rb` produces
keyword argument deprecation warnings in Ruby 2.7.

`ruby2_keywords` is added to `Sequencer::OperationWrapper::method_missing`
function to fix the warning.

Sources:
https://eregon.me/blog/2021/02/13/correct-delegation-in-ruby-2-27-3.html
https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html

gliavallo commented:
@indrekj @bautrey37 @oskarkook
Please review. Help is needed to address the current problems.

indrekj commented:
I think the version should be brought up to 2.7 which is the last 2.x version that is still maintained by ruby team. Please also add github actions support and automate running tests against 2.7, 3.0, 3.1 and 3.2

oskarkook commented:
> Should we add ruby2_keywords dependency somewhere (it worked without it when imported gem)

Yes.

> Should minimal version of Ruby bumped?

I agree that we should bump it to 2.7. Bundler can be updated as well.

oskarkook commented:
To fix tests on 2.7, I'd make the following changes:


https://github.com/pzol/deterministic/blob/master/spec/lib/deterministic/protocol_spec.rb#L34:
```ruby
expect(described_class.constants).to contain_exactly(:Protocol, :Int, :String)
```

https://github.com/pzol/deterministic/blob/master/spec/lib/enum_spec.rb#L73:
```ruby
expect(MyEnym.constants).to contain_exactly(:Nullary, :Unary, :Binary, :Matcher)
```

https://github.com/pzol/deterministic/blob/master/lib/deterministic/sequencer.rb#L1:
```ruby
require 'delegate'
```

bautrey37 commented:
There is an open question on whether ruby2_keywords needs to be added as a dependency. https://github.com/ruby/ruby2_keywords/issues/18

I don't have a reason for or against, I just saw it was a dependency in another place and thought maybe it should be here? But I suppose it will also work if it's not a dependency.

gliavallo committed:
Make tests run under Ruby 2.7+

Tests did not run under Ruby 2.7+. Made code compatible and removed
some deprecation warnings.

Tested with Ruby 2.7 and 3.1.

gliavallo committed:
Update bundler major version to 2

Did not specify minor version to provide compatibility with different
Ruby versions.

gliavallo committed:
Set minimal Ruby version to 2.7

The oldest version of Ruby that is supported right now is 2.7.
Setting the minimal required version to 2.7 here too.

bautrey37 commented:
> ```ruby
> require 'delegate'
> ```

Why is this needed? Just curious.

oskarkook commented:
https://stackoverflow.com/a/60468620

gliavallo committed:
Set up automatic tests

Set up automatic tests to run on every push and pull request.
We use GitHub actions for that purpose.
The testing platform is Ubuntu Linux. The following Ruby versions are
covered: 2.7, 3.0, 3.1, 3.2.

The following resources were used to create workflow file:
1. https://github.com/actions/starter-workflows/blob/main/ci/ruby.yml
2. https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-ruby
3. https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions

gliavallo commented:
@indrekj @bautrey37 @oskarkook 
Made proposed changes and added automated tests by GitHub actions.

gliavallo commented:
Tests are all green:
https://github.com/gliavallo/deterministic/actions/runs/4445752795

oskarkook commented:
Looks good! 👍 `ruby2_keywords` gem doesn't seem to be necessary, as we are using 2.7+ Ruby now and `ruby2_keywords` method is defined in the core library for all the new Ruby versions. The tests pass as well, so I think it's good! Approved from my side :).

indrekj commented on deterministic.gemspec:
```diff
@@ -19,7 +19,7 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
-  spec.add_development_dependency "bundler", "~> 1.3"
+  spec.add_development_dependency "bundler", '~> 2'
```
You're current version basically means `>= 2`. This should be `~> 2.0` which means `>= 2.0` and `< 3.0`

indrekj replied:
see: https://stackoverflow.com/questions/5170547/what-does-tilde-greater-than-mean-in-ruby-gem-dependencies

gliavallo replied:
Fixed.

bautrey37 reviewed (approved):
Thanks for doing this work! 

Let me know when the last comment is fixed so that I can push it to master.

gliavallo commented:
!squash

gliavallo commented:
@bautrey37 It's done now.
